source/common/config/xds_mux/delta_subscription_state.cc

```diff
@@ -132,25 +232,30 @@ DeltaSubscriptionState::getNextRequestInternal() {
   request->set_type_url(typeUrl());
   if (!any_request_sent_yet_in_current_stream_) {
     any_request_sent_yet_in_current_stream_ = true;
+    const bool is_legacy_wildcard = isInitialRequestForLegacyWildcard();
     // initial_resource_versions "must be populated for first request in a stream".
     // Also, since this might be a new server, we must explicitly state *all* of our subscription
     // interest.
-    for (auto const& [resource_name, resource_state] : resource_state_) {
+    for (auto const& [resource_name, resource_state] : requested_resource_state_) {
       // Populate initial_resource_versions with the resource versions we currently have.
       // Resources we are interested in, but are still waiting to get any version of from the
       // server, do not belong in initial_resource_versions. (But do belong in new subscriptions!)
       if (!resource_state.isWaitingForServer()) {
         (*request->mutable_initial_resource_versions())[resource_name] = resource_state.version();
       }
-      // As mentioned above, fill resource_names_subscribe with everything, including names we
-      // have yet to receive any resource for unless this is a wildcard subscription, for which
-      // the first request on a stream must be without any resource names.
-      if (!wildcard_) {
-        names_added_.insert(resource_name);
-      }
+      // We are going over a list of resources that we are interested in, so add them to
+      // resource_names_subscribe.
+      names_added_.insert(resource_name);
+    }
+    for (auto const& [resource_name, resource_version] : wildcard_resource_state_) {
+      (*request->mutable_initial_resource_versions())[resource_name] = resource_version;
     }
-    // Wildcard subscription initial requests must have no resource_names_subscribe.
-    if (wildcard_) {
+    for (auto const& [resource_name, resource_version] : ambiguous_resource_state_) {
+      (*request->mutable_initial_resource_versions())[resource_name] = resource_version;
+    }
+    // If this is a legacy wildcard request, then make sure that the resource_names_subscribe is
+    // empty.
+    if (is_legacy_wildcard) {
       names_added_.clear();
```

Comment:
is a lagacy wildcard stream still considered as a "wildcard stream" after adding some explicit resources? 
e.g. [*, A, B, C]?

Reply:
As said earlier - I suppose it would be clearer to think about wildcard as a resource, not as a stream mode.